File: Repository/UserDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using mpp_proiect_csharp_DianaGliga11.Model;
using mpp_proiect_csharp_DianaGliga11.Repository;

namespace mpp_proiect_csharp_DianaGliga11.Repository
{
    public class UserDBRepository : I_UserDBRepository
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(UserDBRepository));
        private readonly IDictionary<string, string> Props;
        
        public UserDBRepository(IDictionary<string, string> props) 
        {
            log.Info($"{nameof(UserDBRepository)} constructed.");
            Props = props;
        }

        public void Add(User entity)
        {
            log.Info($"Adding User: {entity}");
            var connection = DBUtils.GetConnection(Props);

            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"INSERT INTO Users (username, password) VALUES (@username, @password)";

                var usernameParam = command.CreateParameter();
                usernameParam.ParameterName = "@username";
                usernameParam.Value = entity.UserName;
                command.Parameters.Add(usernameParam);

                var passwordParam = command.CreateParameter();
                passwordParam.ParameterName = "@password";
                passwordParam.Value = entity.Password;
                command.Parameters.Add(passwordParam);

                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while adding new User ", ex);
                throw new EntityRepoException(ex);
            }

            log.Info($"Added User: {entity}");
        }

        public void Remove(long id)
        {
            log.Info($"Removing User: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"DELETE FROM Users WHERE id = @id";
                var idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);
                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while removing User ", ex);
                throw new EntityRepoException(ex);
            }
            
            log.Info($"Removed User: {id}");
        }

        public void Update(long id, User entity)
        {
            log.Info($"Updating User: {entity}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "UPDATE Users SET username = @username, password = @password WHERE id = @id";

                var usernameParam = command.CreateParameter();
                usernameParam.ParameterName = "@username";
                usernameParam.Value = entity.UserName;
                command.Parameters.Add(usernameParam);

                var passwordParam = command.CreateParameter();
                passwordParam.ParameterName = "@password";
                passwordParam.Value = entity.Password;
                command.Parameters.Add(passwordParam);

                var idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);

                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while updating User ", ex);
                throw new EntityRepoException(ex);
            }
            
            log.Info($"Updated User: {entity}");
        }

        public User findById(long id)
        {
            log.Info($"Finding User: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Users WHERE id = @id";
                    var paramID = command.CreateParameter();
                    paramID.ParameterName = "@id";
                    paramID.Value = id;
                    command.Parameters.Add(paramID);

                    using (var dataReader = command.ExecuteReader())
                    {
                        if (dataReader.Read())
                        {
                            User user = Extract(dataReader);
                            log.Info($"Found User: {user}");
                            return user;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error while finding User ", ex);
                throw new EntityRepoException("User was not found");
            }

            return null;
        }

        public IEnumerable<User> getAll()
        {
            log.Info("Getting All Users");
            var connection = DBUtils.GetConnection(Props);
            IList<User> users = new List<User>();
            
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Users";
                
                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        User user = Extract(dataReader);
                        users.Add(user);
                    }
                }
            }
            
            log.Info($"Retrieved {users.Count} users");
            return users;
        }

        public User GetUserByCredentials(string username, string password)
        {
            log.Info($"Searching for User: Username={username}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Users WHERE username = @username AND password = @password";
                    
                    var usernameParam = command.CreateParameter();
                    usernameParam.ParameterName = "@username";
                    usernameParam.Value = username;
                    command.Parameters.Add(usernameParam);

                    var passwordParam = command.CreateParameter();
                    passwordParam.ParameterName = "@password";
                    passwordParam.Value = password;
                    command.Parameters.Add(passwordParam);

                    using (var dataReader = command.ExecuteReader())
                    {
                        if (dataReader.Read())
                        {
                            User user = Extract(dataReader);
                            log.Info($"Found User: {user}");
                            return user;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error while finding User by credentials", ex);
                throw new EntityRepoException("User was not found");
            }

            return null;
        }

        private User Extract(IDataReader dataReader)
        {
            long id = dataReader.GetInt64(0);
            string username = dataReader.GetString(1);
            string password = dataReader.GetString(2);

            User user = new User(username, password)
            {
                Id = id
            };

            return user;
        }
    }
}

File: Repository/src/main/java/org/example/UserDBRepository.java
package org.example;

import Utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.core.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class UserDBRepository implements UserRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = (Logger) LogManager.getLogger();

    public UserDBRepository(Properties props) {
        logger.info("Initializing UserDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(User entity) throws EntityRepoException {
        logger.traceEntry("add user {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "INSERT INTO Users(username, password) VALUES (?, ?)";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getUserName());
            ps.setString(2, entity.getPassword());
            ps.executeUpdate();
            logger.traceExit("User {} added", entity);
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
        logger.traceExit("User {} added", entity);
    }

    @Override
    public void remove(long id) throws EntityRepoException {
        logger.traceEntry("remove user {} ", id);
        Connection connection = dbUtils.getConnection();
        String sql = "DELETE FROM Users WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
            logger.traceExit("User {} removed", id);
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
        logger.traceExit("User {} removed", id);
    }

    @Override
    public List<User> getAll() throws EntityRepoException {
        logger.traceEntry("get all users");
        Connection connection = dbUtils.getConnection();
        List<User> users = new ArrayList<>();
        String sql = "SELECT * FROM Users";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    long id = rs.getInt("id");
                    String username = rs.getString("username");
                    String password = rs.getString("password");
                    User user = new User(username, password);
                    user.setId(id);
                    users.add(user);
                    logger.traceExit("user {} added", user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
        return users;
    }

    @Override
    public User findById(long id) throws EntityRepoException {
        logger.traceEntry("find user by id {}", id);
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Users WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    String username = rs.getString("username");
                    String password = rs.getString("password");
                    User user = new User(username, password);
                    user.setId(id);
                    logger.traceExit("user found {}", user);
                    return user;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
        return null;
    }

    @Override
    public void update(long id, User entity) throws EntityRepoException {
        logger.traceEntry("update user {}", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "UPDATE Users SET username=?, password=? WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getUserName());
            ps.setString(2, entity.getPassword());
            ps.setLong(3, entity.getId());
            ps.executeUpdate();
            logger.traceExit("user {} updated", entity);
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public boolean checkUserPassword(User user) {
        logger.traceEntry("check user credentials for {}", user.getUserName());
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Users WHERE username=? AND password=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, user.getUserName());
            ps.setString(2, user.getPassword());
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next();
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit("check user credentials for {}", user.getUserName());
        return false;
    }

    @Override
    public Optional<User> getByUsernameAndPassword(String username, String password) {
        logger.traceEntry("getByUsernameAndPassword with username {} ", username);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM Users WHERE username=? AND password=?")) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    User user = extract(resultSet);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    private User extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String username = resultSet.getString("username");
        String passwordToken = resultSet.getString("password");

        User user = new User(username, passwordToken);
        user.setId(id);
        user.setPassword(passwordToken);
        return user;
    }

}
